File: js_tools/git_finalize_series.js
return function(args) {
  slop_guard();
  git_assert_clean_workspace("Working tree is dirty. Please commit, stash, or discard changes before finalizing.");

  const canonical_staging = function(branch) {
    if (!branch) return branch;
    const m = String(branch).match(/^(?:slop\/staging\/)+(.*)$/);
    if (!m) return String(branch);
    return "slop/staging/" + m[1];
  };

  const current_branch = git_get_current_branch();
  const target_branch = git_resolve_base_branch(args.target_branch);

  const hash_res = tools.execute_bash({command: "git rev-parse HEAD"});
  const hash = hash_res.stdout.trim();

  const approval_res = tools.query_db({
    sql: "SELECT branch_name, approved_hash FROM patch_approvals WHERE approved_hash = ?",
    params: [hash]
  });

  let approved = false;
  const canonical_current = canonical_staging(current_branch);
  if (approval_res) {
    try {
      const rows = JSON.parse(approval_res);
      for (const row of rows) {
        if (row.approved_hash !== hash) continue;
        if (row.branch_name === current_branch || canonical_staging(row.branch_name) === canonical_current) {
          approved = true;
          break;
        }
      }
    } catch (e) { print("Error parsing JSON from DB: " + e.message); }
  }

  const landed_res = __os_run(`git merge-base --is-ancestor ${shell_escape(hash)} ${shell_escape(target_branch)}`);
  const already_landed = landed_res.exit_code === 0;

  if (!approved && !already_landed) {
    throw new Error("Patch series not approved or hash mismatch. Please obtain approval for hash " + hash + " before finalizing.");
  }

  const res1 = __os_run(`git checkout ${shell_escape(target_branch)}`);
  if (res1.exit_code !== 0) {
    throw new Error("Failed to checkout target branch '" + target_branch + "': " + res1.stderr);
  }

  if (!already_landed) {
    const res2 = __os_run(`git merge --ff-only ${shell_escape(current_branch)}`);
    if (res2.exit_code !== 0) {
      __os_run(`git checkout ${shell_escape(current_branch)}`);
      throw new Error("Merge failed: " + res2.stderr);
    }
  }

  let deleted_staging_branch = false;
  if (current_branch !== target_branch) {
    const del_res = __os_run(`git branch -D ${shell_escape(current_branch)}`);
    deleted_staging_branch = del_res.exit_code === 0;
  }

  tools.query_db({
    sql: "DELETE FROM staging_branches WHERE branch_name = ?",
    params: [current_branch]
  });
  tools.query_db({
    sql: "DELETE FROM staging_branches WHERE branch_name = ?",
    params: [canonical_current]
  });
  tools.query_db({
    sql: "DELETE FROM patch_approvals WHERE branch_name = ?",
    params: [current_branch]
  });
  tools.query_db({
    sql: "DELETE FROM patch_approvals WHERE branch_name = ?",
    params: [canonical_current]
  });

  // Explicitly exit mail mode after a successful finalize.
  // This avoids post-finalize policy confusion when the branch is now main.
  tools.query_db({
    sql: "UPDATE settings SET mode = 'standard' WHERE id = 1"
  });

  const head_res = tools.execute_bash({command: "git rev-parse HEAD"});
  const final_head = (head_res && head_res.stdout) ? head_res.stdout.trim() : "";

  return {
    ok: true,
    action: "finalize_series",
    mail_mode: "off",
    previous_branch: current_branch,
    current_branch: target_branch,
    head: final_head,
    approved: approved,
    already_landed: already_landed,
    merged: !already_landed,
    deleted_staging_branch: deleted_staging_branch,
    cleaned_metadata: true,
    notes: already_landed
      ? ["Patch already landed on target branch", "Cleaned staging metadata", "Mail mode disabled"]
      : ["Series finalized and merged", "Cleaned staging metadata", "Mail mode disabled"]
  };
};

File: core/tools/git_finalize_series.cpp
#include "core/tool_executor.h"

#include "absl/status/status.h"
#include "absl/strings/ascii.h"
#include "absl/strings/str_cat.h"

#include "core/shell_util.h"
#include "core/status_macros.h"
#include "core/tools/common.h"
#include "core/json_utils.h"

namespace slop {
absl::StatusOr<std::string> ToolExecutor::HandleGitFinalizeSeries(
    const nlohmann::json& args, std::shared_ptr<CancellationRequest> cancellation) {
  (void)cancellation;
  RETURN_IF_ERROR(MaybeEnforceMailStagingGuard(mail_mode_));
  RETURN_IF_ERROR(AssertCleanWorkspace());
  if (!db_) return absl::FailedPreconditionError("Database not initialized");

  ASSIGN_OR_RETURN(const std::string current_branch, GetCurrentBranchName());
  ASSIGN_OR_RETURN(const std::string target_branch,
                   ResolveBaseBranch(db_, json_get_or<std::string>(args, "target_branch", "")));
  ASSIGN_OR_RETURN(auto hash_res, RunCommand("git rev-parse HEAD"));
  const std::string hash = std::string(absl::StripAsciiWhitespace(hash_res.stdout_out));

  ASSIGN_OR_RETURN(auto approval_rows_json,
                   db_->Query("SELECT branch_name, approved_hash FROM patch_approvals WHERE approved_hash = ?", {hash}));
  ASSIGN_OR_RETURN(auto approval_rows, ParseDbRows(approval_rows_json, "approval lookup"));
  const std::string canonical_current = CanonicalStagingBranch(current_branch);
  bool approved = false;
  for (const auto& row : approval_rows) {
    if (!row.is_object()) continue;
    const std::string row_hash = row.value("approved_hash", "");
    const std::string row_branch = row.value("branch_name", "");
    if (row_hash == hash && (row_branch == current_branch || CanonicalStagingBranch(row_branch) == canonical_current)) {
      approved = true;
      break;
    }
  }

  ASSIGN_OR_RETURN(auto landed_res,
                   RunCommand(absl::StrCat("git merge-base --is-ancestor ", EscapeShellArg(hash), " ",
                                           EscapeShellArg(target_branch))));
  const bool already_landed = landed_res.exit_code == 0;
  if (!approved && !already_landed) {
    return absl::FailedPreconditionError(
        absl::StrCat("Patch series not approved or hash mismatch. Please obtain approval for hash ", hash,
                     " before finalizing."));
  }

  ASSIGN_OR_RETURN(auto checkout_res, RunCommand(absl::StrCat("git checkout ", EscapeShellArg(target_branch))));
  if (checkout_res.exit_code != 0) {
    return absl::InternalError(
        absl::StrCat("Failed to checkout target branch '", target_branch, "': ", checkout_res.stderr_out));
  }

  if (!already_landed) {
    ASSIGN_OR_RETURN(auto merge_res, RunCommand(absl::StrCat("git merge --ff-only ", EscapeShellArg(current_branch))));
    if (merge_res.exit_code != 0) {
      (void)RunCommand(absl::StrCat("git checkout ", EscapeShellArg(current_branch)));
      return absl::InternalError(absl::StrCat("Merge failed: ", merge_res.stderr_out));
    }
  }

  bool deleted_staging_branch = false;
  if (current_branch != target_branch) {
    ASSIGN_OR_RETURN(auto del_res, RunCommand(absl::StrCat("git branch -D ", EscapeShellArg(current_branch))));
    deleted_staging_branch = del_res.exit_code == 0;
  }

  RETURN_IF_ERROR(db_->Query("DELETE FROM staging_branches WHERE branch_name = ?", {current_branch}).status());
  RETURN_IF_ERROR(db_->Query("DELETE FROM staging_branches WHERE branch_name = ?", {canonical_current}).status());
  RETURN_IF_ERROR(db_->Query("DELETE FROM patch_approvals WHERE branch_name = ?", {current_branch}).status());
  RETURN_IF_ERROR(db_->Query("DELETE FROM patch_approvals WHERE branch_name = ?", {canonical_current}).status());
  RETURN_IF_ERROR(db_->Query("UPDATE settings SET mode = 'standard' WHERE id = 1", {}).status());
  mail_mode_ = false;

  ASSIGN_OR_RETURN(auto final_head_res, RunCommand("git rev-parse HEAD"));
  const std::string final_head = std::string(absl::StripAsciiWhitespace(final_head_res.stdout_out));

  return nlohmann::json({{"ok", true},
                         {"action", "finalize_series"},
                         {"mail_mode", "off"},
                         {"previous_branch", current_branch},
                         {"current_branch", target_branch},
                         {"head", final_head},
                         {"approved", approved},
                         {"already_landed", already_landed},
                         {"merged", !already_landed},
                         {"deleted_staging_branch", deleted_staging_branch},
                         {"cleaned_metadata", true},
                         {"notes",
                          already_landed
                              ? nlohmann::json::array({"Patch already landed on target branch", "Cleaned staging metadata",
                                                       "Mail mode disabled"})
                              : nlohmann::json::array({"Series finalized and merged", "Cleaned staging metadata",
                                                       "Mail mode disabled"})}})
      .dump();
}

}  // namespace slop
